plugins/modules/ec2_vol.py

```diff
@@ -379,125 +385,199 @@ def attach_volume(module, ec2, volume, instance):
+    mapped_block_device = None
+    if not instance_dict:
+        return mapped_block_device
+    if not device_name:
+        return mapped_block_device
+
+    for device in instance_dict.get('block_device_mappings', []):
+        if device['device_name'] == device_name:
+            mapped_block_device = device
+            break
+
+    return mapped_block_device
+
+
+def ensure_tags(module, connection, res_id, res_type, tags, add_only):
+    changed = False
+
+    filters = ansible_dict_to_boto3_filter_list({'resource-id': res_id, 'resource-type': res_type})
+    cur_tags = None
+    try:
+        cur_tags = connection.describe_tags(Filters=filters)
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(e, msg="Couldn't describe tags")
+
+    purge_tags = bool(not add_only)
+    to_update, to_delete = compare_aws_tags(boto3_tag_list_to_ansible_dict(cur_tags.get('Tags')), tags, purge_tags) #TODO: https://github.com/ansible/ansible/pull/45500/files#r218859482
+    final_tags = boto3_tag_list_to_ansible_dict(cur_tags.get('Tags'))
+
+    if to_update:
+        try:
+            if module.check_mode:
+                # update tags
+                final_tags.update(to_update)
+            else:
+                AWSRetry.exponential_backoff()(connection.create_tags)(
+                    Resources=[res_id],
+                    Tags=ansible_dict_to_boto3_tag_list(to_update)
+                )
+
+            changed = True
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e, msg="Couldn't create tags")
+
+    if to_delete:
+        try:
+            if module.check_mode:
+                # update tags
+                for key in to_delete:
+                    del final_tags[key]
+            else:
+                tags_list = []
+                for key in to_delete:
+                    tags_list.append({'Key': key})
+
+                AWSRetry.exponential_backoff()(connection.delete_tags)(Resources=[res_id], Tags=tags_list)
+
+            changed = True
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e, msg="Couldn't delete tags")
+
+    if not module.check_mode and (to_update or to_delete):
+        try:
+            response = connection.describe_tags(Filters=filters)
+            final_tags = boto3_tag_list_to_ansible_dict(response.get('Tags'))
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            module.fail_json_aws(e, msg="Couldn't describe tags")
+
+    return final_tags, changed
+
 
 def main():
-    argument_spec = dict(
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(dict(
         instance=dict(),
         id=dict(),
         name=dict(),
-        volume_size=dict(type='int'),
+        volume_size=dict(),
```

Comment:
It's better to define the 'type' here id possible.

Reply:
I added `type='int'` back into the dictionary.